### PR TITLE
feat: Encrypt documents sent to EDV

### LINKED ISSUE #61: Encrypt documents sent to EDV
Using Aries legacy JWE for now.

### PR DESCRIPTION
**Pull request by DRK3**
- Many of the operations unit tests were using a mocked KMS when they could have used a real one - those have been updated to use real in-memory KMSs now.
- Added shorthand flags for the Bloc domain, external URL host, and universal resolver URL flags.
- Some other minor refactoring and fixes

Signed-off-by: Derek Trider <Derek.Trider@securekey.com>

closes #61

### CODE REVIEW

File: cmd/vc-rest/startcmd/start.go [hunk 1]
```diff
@@ -125,7 +125,7 @@ func createStartCmd(srv server) *cobra.Command {
 			}
 
 			universalResolverURL, err := cmdutils.GetUserSetVar(cmd, universalResolverURLFlagName,
-				universalResolverURLEnvKey, true)
+				universalResolverURLEnvKey, false)
```
Comment (DRK3) [inline cmd/vc-rest/startcmd/start.go]: I noticed that this was set to optional, but it seems to be required since I couldn't go through the REST flow without having a universal resolver running, so I set it to required (isOptional=false). @fqutishat  @sandrask Let me know if you have any concerns.

File: go.mod [hunk 2]
```diff
@@ -15,6 +15,5 @@ require (
 	github.com/spf13/cobra v0.0.6
 	github.com/stretchr/testify v1.4.0
 	github.com/trustbloc/edge-core v0.1.2
-	github.com/trustbloc/edv v0.1.2
-	github.com/trustbloc/sidetree-core-go v0.1.2-0.20200226223710-852e23076586
```
Comment (DRK3) [inline go.mod]: I ran a `go mod tidy` and it removed that sidetree-core-go line. I guess it was left over from a previous commit.

File: pkg/restapi/vc/operation/operations_test.go [hunk 3]
```diff
@@ -37,14 +43,13 @@ import (
 	"github.com/trustbloc/edge-service/pkg/internal/mock/edv"
 )
 
-const multipleContext = `"@context":["https://www.w3.org/2018/credentials/v1","https://w3id.org/citizenship/v1"]`
-
-const validContext = `"@context":["https://www.w3.org/2018/credentials/v1"]`
-
-const invalidContext = `"@context":"https://www.w3.org/2018/credentials/v1"`
+const (
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations_test.go]: I took all the different `const` declarations and grouped them all, so that's why you'll see so many line changes around here.

File: pkg/restapi/vc/operation/operations_test.go [hunk 4]
```diff
@@ -195,9 +184,60 @@ const validVC = `{` +
...
+		validContext + `,
+  "id": "http://example.edu/credentials/1872",
+  "type": "VerifiableCredential",
+  "credentialSubject": {
+    "id": "did:example:ebfeb1f712ebc6f1c276e12ec21"
+  },
+  "issuanceDate": "2010-01-01T19:23:24Z"
+}`
 
-//nolint:lll
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations_test.go]: I noticed there were some super long lines in some of these consts, so I split them up into multiple lines. This way we don't need that `nolint`

File: pkg/restapi/vc/operation/operations_test.go [hunk 5]
```diff
@@ -279,61 +321,12 @@ const inValidVCinVP = `{` +
...
-// VC without issuer
-const invalidVC = `{` +
-	validContext + `,
-  "id": "http://example.edu/credentials/1872",
-  "type": "VerifiableCredential",
-  "credentialSubject": {
-    "id": "did:example:ebfeb1f712ebc6f1c276e12ec21"
-  },
-  "issuanceDate": "2010-01-01T19:23:24Z"
-}`
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations_test.go]: I moved these up so they would be nearby other related `const`s

File: pkg/restapi/vc/operation/operations_test.go [hunk 6]
```diff
@@ -345,7 +338,7 @@ func TestNew(t *testing.T) {
 	t.Run("test error from open store", func(t *testing.T) {
 		client := edv.NewMockEDVClient("test", nil)
 		op, err := New(&mockstore.Provider{ErrOpenStoreHandle: fmt.Errorf("error open store")},
-			client, &kmsmock.CloseableKMS{}, &vdrimock.MockVDRIRegistry{}, "localhost:8080")
+			client, getTestKMS(t), &vdrimock.MockVDRIRegistry{}, "localhost:8080")
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations_test.go]: I noticed that we were using a mock KMS in a lot of places that didn't need a mock - using a real one works fine. Hence you'll see a bunch of these mock KMS calls replaced with a real one.

File: pkg/restapi/vc/operation/operations_test.go [hunk 7]
```diff
@@ -1089,12 +1081,12 @@ func TestStoreVCHandler(t *testing.T) {
 		rr := httptest.NewRecorder()
 		op.storeVCHandler(rr, req)
 		require.Equal(t, http.StatusBadRequest, rr.Code)
-		require.Equal(t, rr.Body.String(), errVaultNotFound.Error())
+		require.Equal(t, rr.Body.String(), "key is nil")
 	})
-	t.Run("store vc err vault not found", func(t *testing.T) {
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations_test.go]: This test was a duplicate, so I removed it (the diff makes it look like I "replaced" it with another test, but in reality I removed the duplicate and added a test).

File: test/bdd/pkg/vc/vc_steps.go [hunk 8]
```diff
@@ -329,14 +329,14 @@ func (e *Steps) retrieveCredential(profileName string) error {
...
-	unescapedResponse, err := strconv.Unquote(string(respBytes))
-	if err != nil {
-		return err
-	}
-
 	expectedCredential := string(e.bddContext.CreatedCredential)
-	if unescapedResponse != expectedCredential {
-		return expectedStringError(expectedCredential, unescapedResponse)
+	receivedCredential := string(respBytes)
+	// For some reason there's an extra line at the end of the credential returned from the create credential handler.
```
Comment (DRK3) [inline test/bdd/pkg/vc/vc_steps.go]: any idea why this is? Normally it doesn't matter since we're just using the JSON unmarshal functions, but it's still odd.

File: cmd/vc-rest/startcmd/start.go (hunk 1 above)
Comment (fqutishat) [reply cmd/vc-rest/startcmd/start.go]: keep it optional

File: test/bdd/pkg/vc/vc_steps.go [hunk 9]
```diff
@@ -329,14 +329,14 @@ func (e *Steps) retrieveCredential(profileName string) error {
...
+	// By adding a space to the received credential we just received we should be able to directly compare the strings.
+	receivedCredential += `
```
Comment (fqutishat) [inline test/bdd/pkg/vc/vc_steps.go]: ??

File: cmd/vc-rest/startcmd/start_test.go [hunk 10]
```diff
@@ -160,6 +162,9 @@ func TestStartCmdValidArgsEnvVar(t *testing.T) {
 	err = os.Setenv(blocDomainEnvKey, "domain")
 	require.Nil(t, err)
 
+	err = os.Setenv(universalResolverURLEnvKey, "localhost:8083")
```
Comment (Baha-sk) [inline cmd/vc-rest/startcmd/start_test.go]: it's a good idea to defer the removal of these test environment variables

File: test/bdd/pkg/vc/vc_steps.go (hunk 8 above)
Comment (Baha-sk) [reply test/bdd/pkg/vc/vc_steps.go]: Instead of appending a breakline, compare against trimmed expectedCredential, but definitely this needs to be investigated.

File: pkg/restapi/vc/operation/operations.go [hunk 11]
```diff
@@ -322,29 +346,65 @@ func (o *Operation) storeVCHandler(rw http.ResponseWriter, req *http.Request) {
...
 
 		return
 	}
 }
 
+func (o *Operation) buildStructuredDoc(data *StoreVCRequest,
+	vc *verifiable.Credential) (*operation.StructuredDocument, error) {
+	// We need an EDV-compliant ID. The UUID that's in the VC is a good fit, since it's 128 bits long already.
+	// Let's convert it to base58 and use it as the StructuredDocument ID.
+	edvDocID, err := convertVCIDToEDVCompatibleFormat(vc.ID)
```
Comment (troyronda) [inline pkg/restapi/vc/operation/operations.go]: I’m concerned about this. Is this identifier stored encrypted or unencrypted?

Comment (DRK3) [reply pkg/restapi/vc/operation/operations.go]: This identifier is not encrypted. Let me know your thoughts on this.

File: test/bdd/pkg/vc/vc_steps.go (hunk 8 above)
Comment (DRK3) [reply test/bdd/pkg/vc/vc_steps.go]: Done!

File: cmd/vc-rest/startcmd/start_test.go (hunk 10 above)
Comment (DRK3) [reply cmd/vc-rest/startcmd/start_test.go]: Done!

File: cmd/vc-rest/startcmd/start.go (hunk 1 above)
Comment (DRK3) [reply cmd/vc-rest/startcmd/start.go]: I can't reproduce that original issue I was having, so I'll make it optional again

Comment (DRK3) [reply cmd/vc-rest/startcmd/start.go]: Done

File: test/bdd/pkg/vc/vc_steps.go (hunk 9 above)
Comment (DRK3) [reply test/bdd/pkg/vc/vc_steps.go]: @fqutishat To clarify: the extra line that I noticed is unrelated to my change, so I created a followup story to investigate and fix it in order to avoid scope creep here.

https://github.com/trustbloc/edge-service/issues/104

Note however that the extra line may be coming from Aries-Framework-Go, since we call into that code while creating a credential. So I (or someone else) might have to make a fix there first.

In the meantime, I've replaced this with the JSON trimming suggestion per @Baha-sk.

File: cmd/vc-rest/startcmd/start_test.go [hunk 12]
```diff
@@ -215,6 +196,41 @@ func TestCreateKMS(t *testing.T) {
 	})
 }
 
+func setEnvVars(t *testing.T) {
+	err := os.Setenv(hostURLEnvKey, "localhost:8080")
```
Comment (Baha-sk) [inline cmd/vc-rest/startcmd/start_test.go]: also backup the original env value and reset it in unsetEnvVars() if it exists

Comment (DRK3) [reply cmd/vc-rest/startcmd/start_test.go]: Talked to @Baha-sk, the env vars seem to get wiped after tests anyway, so this isn't needed

File: pkg/restapi/vc/operation/operations.go (hunk 11 above)
Comment (troyronda) [reply pkg/restapi/vc/operation/operations.go]: Cannot use the credential ID in that case.

Comment (DRK3) [reply pkg/restapi/vc/operation/operations.go]: @sandrask proposed a potential solution. We can hash the VC id and use that instead. This way, you can't derive the VC ID from the EDV document ID. However, if you know the VC ID then you can still retrieve the document. And this way I don't have to store a mapping of VC IDs to EDV document IDs somewhere. 

I've updated the PR with hashing. Let me know what you think. @troyronda @sandrask @Baha-sk

Comment (DRK3) [reply pkg/restapi/vc/operation/operations.go]: (scroll down to the convertVCIDToEDVCompatibleFormat function)

File: cmd/vc-rest/startcmd/start.go [hunk 13]
```diff
@@ -30,31 +30,44 @@ import (
...
-	edvURLFlagShorthand      = "e"
-	edvURLFlagUsage          = "URL EDV instance is running on. Format: HostName:Port."
-	edvURLEnvKey             = "EDV_REST_HOST_URL"
-	blocDomainFlagName       = "bloc-domain"
-	blocDomainFlagUsage      = "Bloc domain"
-	blocDomainEnvKey         = "BLOC_DOMAIN"
-	hostURLExternalFlagName  = "host-url-external"
-	hostURLExternalEnvKey    = "VC_REST_HOST_URL_EXTERNAL"
-	hostURLExternalFlagUsage = "Host External Name:Port This is the URL for the host server as seen externally." +
+	hostURLFlagName      = "host-url"
```
Comment (DRK3) [inline cmd/vc-rest/startcmd/start.go]: A few of these flags were missing their shorthand equivalents. I added them.

I also noticed that there was no spacing between any of these, so I grouped them together based on their purpose.

File: pkg/internal/mock/edv/client.go [hunk 14]
```diff
@@ -12,12 +12,12 @@ import (
...
 type Client struct {
 	edvServerURL            string
-	readDocumentReturnValue []byte
+	ReadDocumentReturnValue *operation.EncryptedDocument
 }
 
 // NewMockEDVClient is the mock version of edv client
-func NewMockEDVClient(edvServerURL string, readDocumentReturnValue []byte) *Client {
-	return &Client{edvServerURL: edvServerURL, readDocumentReturnValue: readDocumentReturnValue}
+func NewMockEDVClient(edvServerURL string, readDocumentReturnValue *operation.EncryptedDocument) *Client {
```
Comment (DRK3) [inline pkg/internal/mock/edv/client.go]: This is due to a change in the interface defined in the EDV dependency (which was updated in this PR).

File: cmd/vc-rest/startcmd/start.go [hunk 15]
```diff
@@ -103,46 +117,12 @@ func createStartCmd(srv server) *cobra.Command {
...
-			}
-
-			universalResolverURL, err := cmdutils.GetUserSetVar(cmd, universalResolverURLFlagName,
-				universalResolverURLEnvKey, true)
-			if err != nil {
-				return err
-			}
-
-			mode, err := cmdutils.GetUserSetVar(cmd, modeFlagName, modeEnvKey, true)
+			hostURL, edvURL, blocDomain, hostURLExternal, universalResolverURL, mode, localStorePath, err :=
```
Comment (Baha-sk) [inline cmd/vc-rest/startcmd/start.go]: consider creating a struct with the returned fields and return this struct instead

File: cmd/vc-rest/startcmd/start.go [hunk 16]
```diff
@@ -151,19 +131,72 @@ func createStartCmd(srv server) *cobra.Command {
...
 				mode:                 mode,
+				localStorePath:       localStorePath,
 			}
+
 			return startEdgeService(parameters)
 		},
 	}
 }
 
+func getVCRestParameters(cmd *cobra.Command) (string, string, string, string, string, string, string, error) {
```
Comment (Baha-sk) [inline cmd/vc-rest/startcmd/start.go]: returning more than 2/3 values is a sign they should be grouped in a struct instead.

File: pkg/restapi/vc/operation/operations.go [hunk 17]
```diff
@@ -73,19 +80,27 @@ type vcStatusManager interface {
...
 type Client interface {
 	CreateDataVault(config *operation.DataVaultConfiguration) (string, error)
-	CreateDocument(vaultID string, document *operation.StructuredDocument) (string, error)
-	ReadDocument(vaultID, docID string) ([]byte, error)
+	CreateDocument(vaultID string, document *operation.EncryptedDocument) (string, error)
+	ReadDocument(vaultID, docID string) (*operation.EncryptedDocument, error)
+}
+
+type cryptoProvider struct {
+	crypto legacykms.KeyManager
```
Comment (Baha-sk) [inline pkg/restapi/vc/operation/operations.go]: do not confuse crypto with kms.. they are two different concepts.. KeyManager is a kms, this should be a kmsProvider with a kms field instead of cryptoProvider with crypto field

File: pkg/restapi/vc/operation/operations.go [hunk 18]
```diff
@@ -97,12 +112,33 @@ func New(provider storage.Provider, client Client, kms legacykms.KMS, vdri vdria
 		return nil, fmt.Errorf("failed to instantiate new csl status: %w", err)
 	}
 
+	cryptoProv := cryptoProvider{
+		crypto: kms,
```
Comment (Baha-sk) [inline pkg/restapi/vc/operation/operations.go]: it is more clear when used here... a kms is not a crypto instance

File: cmd/vc-rest/startcmd/start.go [hunk 19]
```diff
@@ -30,31 +30,44 @@ import (
...
+	universalResolverURLFlagUsage     = "Universal Resolver instance is running on. Format: HostName:Port."
+	universalResolverURLEnvKey        = "UNIVERSAL_RESOLVER_HOST_URL"
+
+	modeFlagName      = "mode"
+	modeFlagShorthand = "m"
+	modeFlagUsage     = "Mode in which the vc-rest service will run. Possible values: " +
 		"['issuer', 'verifier'] (default: issuer)."
 	modeEnvKey = "VC_REST_MODE"
 
+	localStorePathFlagName      = "local-store-path"
```
Comment (fqutishat) [inline cmd/vc-rest/startcmd/start.go]: what this flag mean?

File: pkg/restapi/vc/operation/operations.go [hunk 20]
```diff
@@ -322,41 +361,90 @@ func (o *Operation) storeVCHandler(rw http.ResponseWriter, req *http.Request) {
...
-func convertVCIDToEDVCompatibleFormat(vcID string) (string, error) {
-	vcIDParts := strings.Split(vcID, "/")
+func (o *Operation) buildStructuredDoc(data *StoreVCRequest,
+	vc *verifiable.Credential) (*operation.StructuredDocument, error) {
+	var edvDocID string
 
-	uuidFromVCID := vcIDParts[len(vcIDParts)-1]
+	idFromMapping, err := o.idMapping.Get(vc.ID)
+	if err != nil {
+		edvDocID, err = generateEDVCompatibleID()
```
Comment (Baha-sk) [inline pkg/restapi/vc/operation/operations.go]: we lost the original `err` here, are you sure you don't want to wrap/embed the 2 errors.. or at least log a warning with the original one?

since the original err is from a `Get()` function, I would assume you want to call `generateEDVCompatibleID` only if `Get()` returned a `notFound` error and other errors are valid error type that should return automatically .. 

this logic needs to be revisited.

File: test/bdd/fixtures/vc-rest/docker-compose.yml [hunk 21]
```diff
@@ -17,6 +17,7 @@ services:
       - EDV_REST_HOST_URL=${EDV_HOST}:${EDV_PORT}
       - BLOC_DOMAIN=testnet.trustbloc.local:80
       - UNIVERSAL_RESOLVER_HOST_URL=http://uni-resolver-web:8080/1.0/identifiers
+      - LOCAL_STORE_PATH=bdd-test-local-store
```
Comment (fqutishat) [inline test/bdd/fixtures/vc-rest/docker-compose.yml]: ???

File: cmd/vc-rest/startcmd/start.go (hunk 16 above)
Comment (fqutishat) [reply cmd/vc-rest/startcmd/start.go]: you need to change this method

File: pkg/restapi/vc/operation/operations.go [hunk 22]
```diff
@@ -322,41 +361,90 @@ func (o *Operation) storeVCHandler(rw http.ResponseWriter, req *http.Request) {
...
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		edvDocID = string(idFromMapping)
+	}
 
-	parsedUUID, err := uuid.Parse(uuidFromVCID)
+	doc := operation.StructuredDocument{}
+	doc.ID = edvDocID
```
Comment (Baha-sk) [inline pkg/restapi/vc/operation/operations.go]: Shouldn't there be a check here `if edvDocID == ""` in case err was nil but edvDocID returned was empty above ? 

I guess this is another reason this logic needs to be revisited.

File: cmd/vc-rest/startcmd/start.go (hunk 15 above)
Comment (DRK3) [reply cmd/vc-rest/startcmd/start.go]: Oh, yes you're right. In fact, there's already a struct called vcRestParameters that's a perfect fit already. In hindsight it seems obvious...not sure why I missed that before. I've updated the code.

File: cmd/vc-rest/startcmd/start.go (hunk 16 above)
Comment (DRK3) [reply cmd/vc-rest/startcmd/start.go]: Oh, yes you're right. In fact, there's already a struct called vcRestParameters that's a perfect fit already. In hindsight it seems obvious...not sure why I missed that before. I've updated the code.

File: pkg/restapi/vc/operation/operations.go (hunk 17 above)
Comment (DRK3) [reply pkg/restapi/vc/operation/operations.go]: Gotcha. I've updated the code!

File: pkg/restapi/vc/operation/operations.go (hunk 18 above)
Comment (DRK3) [reply pkg/restapi/vc/operation/operations.go]: I see what you mean! I've updated the code.

File: pkg/restapi/vc/operation/operations.go (hunk 20 above)
Comment (DRK3) [reply pkg/restapi/vc/operation/operations.go]: Ah, when I first did this I had falsely assumed that the Aries storage didn't have a generic "not found" error, but now that I take a closer look it does indeed! I've updated the code to check for the specific condition now.

File: pkg/restapi/vc/operation/operations.go (hunk 22 above)
Comment (DRK3) [reply pkg/restapi/vc/operation/operations.go]: I  don't believe it's possible for edvDocID to be blank at this point, unless someone manually put a blank string in the database. But even if it was blank, the EDV would just return an error, which would cause the store VC handler to return an error back to the person who called that endpoint.

File: cmd/vc-rest/startcmd/start.go (hunk 19 above)
Comment (DRK3) [reply cmd/vc-rest/startcmd/start.go]: VCs have IDs and EDV encrypted documents have IDs. When someone calls a REST endpoint in edge-service, they will provide the VC ID, however this will not be the same as the ID of the associated encrypted document that contains the encrypted VC. So I need to map from VC ID -> EDV encrypted document ID, and I need to store that somewhere. So I store that in a local LevelDB database. This flag specifies where the database is created in the file system.

(It's like the [db-path flag in Aries-Framework-Go)](https://github.com/hyperledger/aries-framework-go/blob/6f509cae073a587e086972aaa93b9a878f88dfc2/cmd/aries-agent-rest/startcmd/start.go#L50)

File: test/bdd/fixtures/vc-rest/docker-compose.yml (hunk 21 above)
Comment (DRK3) [reply test/bdd/fixtures/vc-rest/docker-compose.yml]: See my response to your other comment. With that in mind: within the docker container it doesn't really matter where the storage is for these tests, as long as it's writable.

File: pkg/restapi/vc/operation/operations.go [hunk 23]
```diff
@@ -97,12 +112,33 @@ func New(provider storage.Provider, client Client, kms legacykms.KMS, vdri vdria
...
+	}
+
+	packer := authcrypt.New(kmsProv)
+
+	_, senderKey, err := kms.CreateKeySet()
+	if err != nil {
+		return nil, err
+	}
+
+	localStoreProvider := leveldb.NewProvider(localStorePath)
```
Comment (fqutishat) [inline pkg/restapi/vc/operation/operations.go]: use provider.CreateStore instead of leveldb

Comment (fqutishat) [reply pkg/restapi/vc/operation/operations.go]: similar to this err := provider.CreateStore(credentialStoreName)

Comment (fqutishat) [reply pkg/restapi/vc/operation/operations.go]: no need for localStorePath just use localStoreName

File: test/bdd/fixtures/vc-rest/docker-compose.yml (hunk 21 above)
Comment (fqutishat) [reply test/bdd/fixtures/vc-rest/docker-compose.yml]: no need

File: pkg/restapi/vc/operation/operations.go [hunk 24]
```diff
@@ -20,11 +21,14 @@ import (
...
 	"github.com/google/uuid"
 	"github.com/gorilla/mux"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/packer/legacy/authcrypt"
 	"github.com/hyperledger/aries-framework-go/pkg/doc/did"
 	"github.com/hyperledger/aries-framework-go/pkg/doc/signature/ed25519signature2018"
 	"github.com/hyperledger/aries-framework-go/pkg/doc/verifiable"
 	vdriapi "github.com/hyperledger/aries-framework-go/pkg/framework/aries/api/vdri"
 	"github.com/hyperledger/aries-framework-go/pkg/kms/legacykms"
+	ariesstorage "github.com/hyperledger/aries-framework-go/pkg/storage"
+	"github.com/hyperledger/aries-framework-go/pkg/storage/leveldb"
```
Comment (fqutishat) [inline pkg/restapi/vc/operation/operations.go]: you don't need to use ariesstorage use github.com/trustbloc/edge-core/pkg/storage

File: cmd/vc-rest/startcmd/start.go (hunk 19 above)
Comment (fqutishat) [reply cmd/vc-rest/startcmd/start.go]: no need to use LOCAL_STORE_PATH follow credential store way

**Comment by fqutishat** [comment]
@Baha-sk are you ok using ariesstorage????

File: cmd/vc-rest/startcmd/start_test.go [hunk 25]
```diff
@@ -215,6 +195,41 @@ func TestCreateKMS(t *testing.T) {
...
+	require.NoError(t, err)
+
+	err = os.Unsetenv(edvURLEnvKey)
+	require.NoError(t, err)
+
+	err = os.Unsetenv(blocDomainEnvKey)
+	require.NoError(t, err)
+}
+
+func checkFlagPropertiesCorrect(t *testing.T, cmd *cobra.Command, flagName, flagShorthand, flagUsage string) {
```
Comment (DRK3) [inline cmd/vc-rest/startcmd/start_test.go]: I just moved this method down, so all the Test methods could be together.

File: pkg/restapi/vc/operation/operations.go [hunk 26]
```diff
@@ -130,17 +170,20 @@ type Operation struct {
...
 }
 
 func (o *Operation) createCredentialHandler(rw http.ResponseWriter, req *http.Request) {
 	data := CreateCredentialRequest{}
-	err := json.NewDecoder(req.Body).Decode(&data)
 
+	err := json.NewDecoder(req.Body).Decode(&data)
 	if err != nil {
-		o.writeErrorResponse(rw, http.StatusBadRequest, "Failed to write response for invalid request received")
+		o.writeErrorResponse(rw, http.StatusBadRequest, invalidRequestErrMsg)
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations.go]: This error message was incorrect, so I fixed it.

File: pkg/restapi/vc/operation/operations.go [hunk 27]
```diff
@@ -281,7 +324,7 @@ func (o *Operation) createProfileHandler(rw http.ResponseWriter, req *http.Reque
 	err := json.NewDecoder(req.Body).Decode(&data)
 
 	if err != nil {
-		o.writeErrorResponse(rw, http.StatusBadRequest, "Failed to write response for invalid request received")
+		o.writeErrorResponse(rw, http.StatusBadRequest, invalidRequestErrMsg)
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations.go]: This error message was incorrect, so I fixed it.

File: pkg/restapi/vc/operation/operations.go [hunk 28]
```diff
@@ -130,17 +170,20 @@ type Operation struct {
...
 }
 
 func (o *Operation) createCredentialHandler(rw http.ResponseWriter, req *http.Request) {
 	data := CreateCredentialRequest{}
-	err := json.NewDecoder(req.Body).Decode(&data)
 
+	err := json.NewDecoder(req.Body).Decode(&data)
 	if err != nil {
-		o.writeErrorResponse(rw, http.StatusBadRequest, "Failed to write response for invalid request received")
+		o.writeErrorResponse(rw, http.StatusBadRequest, fmt.Sprintf(invalidRequestErrMsg+": %s", err.Error()))
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations.go]: This error message was incorrect, so I fixed it.

File: pkg/restapi/vc/operation/operations.go [hunk 29]
```diff
@@ -278,16 +321,15 @@ func (o *Operation) updateCredentialStatusHandler(rw http.ResponseWriter, req *h
 
 func (o *Operation) createProfileHandler(rw http.ResponseWriter, req *http.Request) {
 	data := ProfileRequest{}
-	err := json.NewDecoder(req.Body).Decode(&data)
 
+	err := json.NewDecoder(req.Body).Decode(&data)
 	if err != nil {
-		o.writeErrorResponse(rw, http.StatusBadRequest, "Failed to write response for invalid request received")
+		o.writeErrorResponse(rw, http.StatusBadRequest, fmt.Sprintf(invalidRequestErrMsg+": %s", err.Error()))
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations.go]: This error message was incorrect, so I fixed it.

File: pkg/restapi/vc/operation/operations.go [hunk 30]
```diff
@@ -322,7 +364,7 @@ func (o *Operation) storeVCHandler(rw http.ResponseWriter, req *http.Request) {
 
 	err := json.NewDecoder(req.Body).Decode(&data)
 	if err != nil {
-		o.writeErrorResponse(rw, http.StatusBadRequest, "invalid request received")
+		o.writeErrorResponse(rw, http.StatusBadRequest, fmt.Sprintf(invalidRequestErrMsg+": %s", err.Error()))
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations.go]: This error message was incorrect, so I fixed it.

File: pkg/restapi/vc/operation/operations.go (hunk 20 above)
Comment (DRK3) [reply pkg/restapi/vc/operation/operations.go]: Just as a heads up: this now uses the Value Not Found error from edge-core. Same idea. Works the same way.

File: pkg/restapi/vc/operation/operations_test.go [hunk 31]
```diff
@@ -20,11 +20,16 @@ import (
...
 
+	"github.com/btcsuite/btcutil/base58"
+
 	"github.com/gorilla/mux"
 	"github.com/hyperledger/aries-framework-go/pkg/doc/did"
 	"github.com/hyperledger/aries-framework-go/pkg/doc/verifiable"
+	"github.com/hyperledger/aries-framework-go/pkg/kms/legacykms"
 	kmsmock "github.com/hyperledger/aries-framework-go/pkg/mock/kms/legacykms"
 	vdrimock "github.com/hyperledger/aries-framework-go/pkg/mock/vdri"
+	ariesstorage "github.com/hyperledger/aries-framework-go/pkg/storage"
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations_test.go]: Aries storage is used here for manipulating KMS objects which exist in some tests. I believe the edge-core storage technically has the same signature, so they might appear to be compatible interfaces but the semantics are different due to the separate of Open/Create store in the edge-core version.

File: pkg/restapi/vc/operation/operations_test.go [hunk 32]
```diff
@@ -1573,3 +1709,34 @@ func (m *mockVCStatusManager) UpdateVCStatus(v *verifiable.Credential, profile *
 func (m *mockVCStatusManager) GetCSL(id string) (*cslstatus.CSL, error) {
 	return m.getCSLValue, m.getCSLErr
 }
+
+// FailOnSecondOpenMockProvider is a mock provider that returns an err the second time the OpenStore method is called.
+type FailOnSecondOpenMockProvider struct {
```
Comment (DRK3) [inline pkg/restapi/vc/operation/operations_test.go]: Interestingly enough, I actually have a case where I needed it to fail the second time...  (I remember I asked about this hypothetical situation awhile back @sandrask, but I didn't think I'd actually see it so soon)
